include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+		result.w = pdf;
+	} CASE_END
+#endif
+#ifdef OP_CONDUCTOR
+	CASE_BEGIN(op, OP_CONDUCTOR) {
+		result.xyz = instr_execute_cos_eval_CONDUCTOR(instr, s, uf, bxdf_eval_scalar_part, params, bsdf_data);
+		result.w = pdf;
+	} CASE_END
+#endif
+#ifdef OP_PLASTIC
+	CASE_BEGIN(op, OP_PLASTIC) {
+		result = instr_execute_cos_eval_pdf_PLASTIC(instr, s, uf, regs, bxdf_eval_scalar_part, params, bsdf_data, pdf);
+	} CASE_END
+#endif
+#ifdef OP_COATING
+	CASE_BEGIN(op, OP_COATING) {
+		result.xyz = instr_execute_cos_eval_COATING(instr, srcs, params, bsdf_data);
+		result.w = pdf;
+	} CASE_END
+#endif
+#ifdef OP_DIELECTRIC
+	CASE_BEGIN(op, OP_DIELECTRIC) {
+		result.xyz = instr_execute_cos_eval_DIELECTRIC(instr, s, bxdf_eval_scalar_part);
+		result.w = pdf;
+	} CASE_END
+#endif
+#ifdef OP_THINDIELECTRIC
+	CASE_BEGIN(op, OP_THINDIELECTRIC) {
+		result = instr_execute_cos_eval_pdf_THINDIELECTRIC(instr, s, regs, params, bsdf_data);
+	} CASE_END
+#endif
+#ifdef OP_BLEND
+	CASE_BEGIN(op, OP_BLEND) {
+		result = instr_execute_cos_eval_pdf_BLEND(instr, srcs, params, bsdf_data);
+	} CASE_END
+#endif
+#ifdef OP_BUMPMAP
+	CASE_BEGIN(op, OP_BUMPMAP) {
+		instr_execute_BUMPMAP(instr, srcs);
+	} CASE_END
+#endif
+#ifdef OP_SET_GEOM_NORMAL
+	CASE_BEGIN(op, OP_SET_GEOM_NORMAL) {
+		instr_execute_SET_GEOM_NORMAL();
+	} CASE_END
+#endif
+	CASE_OTHERWISE
+	{} //else "empty braces"
+	END_CASES
+
+	if (op_isBXDForBlend(op))
+		writeReg(REG_DST(regs), result);
+}
+
+eval_and_pdf_t irr_bsdf_eval_and_pdf(in instr_stream_t stream, inout irr_glsl_LightSample s, in instr_t generator, inout irr_glsl_AnisotropicMicrofacetCache uf)
+{
+	bool ts = false;
+	const bool is_generator_bsdf = op_isBRDF(instr_getOpcode(generator));
+	instr_execute_SET_GEOM_NORMAL();
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+		uint op = instr_getOpcode(instr);
+
+		if (INSTR_1ST_DWORD(instr) != INSTR_1ST_DWORD(generator))
+		{
+			instr_eval_and_pdf_execute(instr, s, uf, ts);
+		}
+		else
+		{
+			//assert(isBXDF(instr));
+			uvec3 regs = instr_decodeRegisters(instr);
+			writeReg(REG_DST(regs), eval_and_pdf_t(0.0));
+		}
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			s = irr_glsl_createLightSample(s.L, currInteraction);
+			uf = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		}
+#endif
+	}
+
+	eval_and_pdf_t eval_and_pdf = readReg4(0u);
+	return eval_and_pdf;
+}
+
+irr_glsl_AnisotropicMicrofacetCache getSmoothMicrofacetCache(in float NdotV, in float NdotL)
+{
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	const float d = 1e-4;//some delta to avoid zeroes... not a perfect solution but for now..
+	uf.isotropic.VdotH = NdotV-d;
+	uf.isotropic.LdotH = NdotL-d;
+	uf.isotropic.NdotH = d;
+	uf.isotropic.NdotH2 = d*d;
+	uf.TdotH = 1.0-d;
+	uf.BdotH = 1.0-d;
```

Comment:
why would zeros be a problem?

Also `TdotH` and `BdotH` should be equal to 0.0 !

You got everything the wrong way round XD

Reply:
`NdotH = 1.0`

Reply:
this whole function's existence doesnt make sense really

Reply:
should be unneeded if you use the same func to compute the microfacet case for diffuse, difftrans and thindielectric (only in the reflective case obvs)